### PR TITLE
Refactor challenges to use cycle count instead of time

### PR DESCRIPTION
**Pull request by merwanehamadi**
### Background

Continuous agents need to be stopped after a certain time, but we don't know that time in advance.
Plus there could be some rate limit errors that make the agent take more time.
So instead of using time let's use the number of cycle.
Other code cleanup added.
### Changes

use cycle count instead of time
enforce type hinting on challenges
create a test for challenges to be formatted properly (it's a test testing the tests)

### PR Quality Checklist
- [x] My pull request is atomic and focuses on a single change.
- [x] I have thoroughly tested my changes with multiple different prompts.
- [x] I have considered potential risks and mitigations for my changes.
- [x] I have documented my changes clearly and comprehensively.
- [x] I have not snuck in any "extra" small tweaks changes

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -45,6 +45,10 @@ jobs:
         run: isort . --check
         if: success() || failure()
 
+      - name: Check mypy formatting
```
Comment (richbeales) [inline .github/workflows/ci.yml:48]: do we need another linter?

Comment (waynehamadi) [reply .github/workflows/ci.yml:48]: this is just for type hint, this was suggested 2 weeks ago.
Unfortunately black and flake8 don't provide type hint checks.
this only applies to challenges

**Comment by dschonholtz** [comment]
Could you describe exactly what the cycle count is doing?
I like the idea of enforcing mypy.

File: tests/integration/challenges/utils.py [hunk 2]
```diff
@@ -45,19 +48,46 @@ def generate_noise(noise_size) -> str:
...
+
+
+def setup_mock_input(monkeypatch: pytest.MonkeyPatch, cycle_count: int) -> None:
+    """
+    Sets up the mock input for testing.
+
+    :param monkeypatch: pytest's monkeypatch utility for modifying builtins.
+    :param cycle_count: The number of cycles to mock.
+    """
+    input_sequence = ["y"] * (cycle_count) + ["EXIT"]
```
Comment (dschonholtz) [inline tests/integration/challenges/utils.py:76]: Ah you can ignore my previous comment.
Neat.
2 or 3 cycles is enough to reliably finish some challenges? That seems very low.

**Comment by dschonholtz** [comment]
Generally this seems like a good thing

File: tests/integration/challenges/utils.py (hunk 2 above)
Comment (waynehamadi) [reply tests/integration/challenges/utils.py:76]: yeah correct, these challenges are the basic ones.

**Review by ntindle — APPROVED** [review]

**Pull request merged**